core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -55,35 +60,44 @@
    private static final Log log = LogFactory.getLog(ClusterExpirationManager.class);
    private static final boolean trace = log.isTraceEnabled();
 
-   private static final int MAX_ASYNC_EXPIRATIONS = 5;
+   private static final int MAX_ASYNC_EXPIRATIONS = 100;
 
    @Inject protected ComponentRef<AdvancedCache<K, V>> cacheRef;
    @Inject protected ComponentRef<CommandsFactory> cf;
    @Inject protected RpcManager rpcManager;
    @Inject protected DistributionManager distributionManager;
 
    private AdvancedCache<K, V> cache;
+   private Address localAddress;
 
    @Override
    public void start() {
       super.start();
       // Data container entries are retrieved directly, so we don't need to worry about an encodings
       this.cache = AbstractDelegatingCache.unwrapCache(cacheRef.wired()).getAdvancedCache();
+      this.localAddress = cache.getCacheManager().getAddress();
    }
 
    @Override
    public void processExpiration() {
       long start = 0;
+      int removedEntries = 0;
       if (!Thread.currentThread().isInterrupted()) {
          try {
             if (trace) {
                log.trace("Purging data container of expired entries");
                start = timeService.time();
             }
             // We limit it so there is only so many async expiration removals done at the same time
-            List<CompletableFuture> futures = new ArrayList<>(MAX_ASYNC_EXPIRATIONS);
+            Semaphore expirationPermits = new Semaphore(MAX_ASYNC_EXPIRATIONS);
+            Set<CompletableFuture<?>> futures = ConcurrentHashMap.newKeySet();
```

Comment:
IMO it would be simpler if we used an `ArrayBlockingQueue(MAX_ASYNC_EXPIRATIONS)`.
For each expired entry, on `whenComplete` we add the completed future to the queue.
For the first MAX_ASYNC_EXPIRATIONS entries, we start the expiration without a delay, but for the rest we poll a future from the queue and `join()` it before starting a new expiration.

It's probably best to not stop the purge if one expiration fails, like master does, but we should report the number of failed expirations at the end instead of a generic "Caught exception purging data container!".

Reply:
Interesting, I can try something like that.

Reply:
Changed.